Sming/SmingCore/Wire.h

```diff
@@ -1,60 +1,91 @@
- * Created 2015 by Skurydin Alexey
- * http://github.com/anakod/Sming
- * All files of the Sming Core are provided under the LGPL v3 license.
- ****/
+/*
+  TwoWire.h - TWI/I2C library for Arduino & Wiring
+  Copyright (c) 2006 Nicholas Zambetti.  All right reserved.
 
-#ifndef _SMING_CORE_WIRE_H_
-#define _SMING_CORE_WIRE_H_
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
 
-#include "../Wiring/WiringFrameworkDependencies.h"
-#include "../Wiring/Stream.h"
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
 
-// Default values
-#define I2C_DEFAULT_SCL_PIN		0
-#define I2C_DEFAULT_SDA_PIN		2
-#define BUFFER_LENGTH			48
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 
-class SoftI2cMaster;
+  Modified 2012 by Todd Krein (todd@krein.org) to implement repeated starts
+  Modified December 2014 by Ivan Grokhotkov (ivan@esp8266.com) - esp8266 support
+  Modified April 2015 by Hrsto Gochkov (ficeto@ficeto.com) - alternative esp8266 support
+*/
+
+#ifndef TwoWire_h
+#define TwoWire_h
+
+#include <inttypes.h>
+#include "Stream.h"
+
+#define BUFFER_LENGTH 32
 
 class TwoWire : public Stream
 {
-public:
-	TwoWire(int pinSCL, int pinSDA);
-	virtual ~TwoWire();
-
-	void pins(int pinSCL, int pinSDA); // Can be called only before begin()
-	void begin();
-
-	void beginTransmission(uint8_t address);
-	uint8_t endTransmission(bool sendStop = true);
-	uint8_t requestFrom(int address, int quantity, bool sendStop = true);
-
-	virtual int available();
-	virtual int read();
-	virtual int peek();
-	virtual void flush();
-	size_t write(uint8_t data);
-	size_t write(const uint8_t *data, size_t quantity);
-
-protected:
-	uint8_t pushData();
-
-private:
-	SoftI2cMaster* master;
-	int SCL;
-	int SDA;
-	int targetAddress;
-
-	uint8_t txBuf[BUFFER_LENGTH];
-	int txLen;
-
-	uint8_t rxBuf[BUFFER_LENGTH];
-	int rxLen;
-	int rxPos;
+  private:
+    static uint8_t rxBuffer[];
+    static uint8_t rxBufferIndex;
+    static uint8_t rxBufferLength;
+
+    static uint8_t txAddress;
+    static uint8_t txBuffer[];
+    static uint8_t txBufferIndex;
+    static uint8_t txBufferLength;
+
+    static uint8_t transmitting;
+    static void (*user_onRequest)(void);
+    static void (*user_onReceive)(int);
+    static void onRequestService(void);
+    static void onReceiveService(uint8_t*, int);
+  public:
+    TwoWire();
+    void begin(int scl, int sda);
+    void pins(int scl, int sda);
```

Comment:
This will require porting for more hw libs as the argument order is reversed - or writing compat layers on top. Keeping compatibility with Arduino will make drop-in replacement for Wire.h. Hard to say what will be better although I don't mind writing some code to make existing lib working.